pandas/core/array_algos/cast.py

```diff
@@ -0,0 +1,117 @@
+import inspect
+
+import numpy as np
+
+from pandas._typing import (
+    ArrayLike,
+    DtypeObj,
+)
+
+from pandas.core.dtypes.cast import (
+    astype_dt64_to_dt64tz,
+    astype_nansafe,
+)
+from pandas.core.dtypes.common import (
+    is_datetime64_dtype,
+    is_datetime64tz_dtype,
+    is_dtype_equal,
+    pandas_dtype,
+)
+from pandas.core.dtypes.dtypes import ExtensionDtype
+
+from pandas.core.arrays import ExtensionArray
+
+
+def astype_array(values: ArrayLike, dtype: DtypeObj, copy: bool = False):
+    """
+    Cast array to the new dtype.
+
+    Parameters
+    ----------
+    values : ndarray or ExtensionArray
+    dtype : dtype object
+    copy : bool, default False
+        copy if indicated
+
+    Returns
+    -------
+    ndarray or ExtensionArray
+    """
+    if (
+        values.dtype.kind in ["m", "M"]
+        and dtype.kind in ["i", "u"]
+        and isinstance(dtype, np.dtype)
+        and dtype.itemsize != 8
+    ):
+        # TODO(2.0) remove special case once deprecation on DTA/TDA is enforced
+        msg = rf"cannot astype a datetimelike from [{values.dtype}] to [{dtype}]"
+        raise TypeError(msg)
+
+    if is_datetime64tz_dtype(dtype) and is_datetime64_dtype(values.dtype):
+        return astype_dt64_to_dt64tz(values, dtype, copy, via_utc=True)
+
+    if is_dtype_equal(values.dtype, dtype):
+        if copy:
+            return values.copy()
+        return values
+
+    if isinstance(values, ExtensionArray):
+        values = values.astype(dtype, copy=copy)
+
+    else:
+        values = astype_nansafe(values, dtype, copy=copy)
+
+    # now in ObjectBlock._maybe_coerce_values(cls, values):
+    if isinstance(dtype, np.dtype) and issubclass(values.dtype.type, str):
+        values = np.array(values, dtype=object)
+
+    return values
+
+
+def astype_array_safe(values, dtype, copy: bool = False, errors: str = "raise"):
```

Comment:
values: ArrayLike; dtype: DtypeObj?

Reply:
mypy isn't smart enough for that (we also didn't annotate `dtype` in the code from where I copied this). But will add `values: ArrayLike`

Reply:
make sense.  if we move the pandas_dtype call up into the caller we can do it once instead of (block|array)-wise

Reply:
>  if we move the pandas_dtype call up into the caller we can do it once instead of (block|array)-wise

Then also the `dtype` validation needs to be moved up (requiring some more changes to avoid duplication of that part). At the moment, it's a rather straightfoward cut and paste from blocks.py to cast.py, so I would maybe prefer to keep it that way for this PR.